File: src/sdk/python/ofa_agent/agent.py
"""
OFA Agent Core
Sprint 29: Python Agent SDK
"""

import asyncio
import json
import logging
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional, Callable, Awaitable

logger = logging.getLogger(__name__)


class AgentState(Enum):
    """Agent状态"""
    INITIALIZING = "initializing"
    CONNECTING = "connecting"
    ONLINE = "online"
    BUSY = "busy"
    OFFLINE = "offline"
    ERROR = "error"


@dataclass
class AgentConfig:
    """Agent配置"""
    agent_id: str = ""
    name: str = "Python Agent"
    center_url: str = "localhost:9090"
    connection_type: str = "grpc"  # grpc, http, websocket
    heartbeat_interval: float = 30.0
    reconnect_interval: float = 5.0
    max_reconnect_attempts: int = 10
    skills: List[str] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)
    tls_enabled: bool = False
    timeout: float = 30.0

    def __post_init__(self):
        if not self.agent_id:
            self.agent_id = f"python-agent-{uuid.uuid4().hex[:8]}"


@dataclass
class AgentInfo:
    """Agent信息"""
    id: str
    name: str
    type: str = "python"
    version: str = "8.1.0"
    platform: str = "python"
    skills: List[str] = field(default_factory=list)
    state: AgentState = AgentState.OFFLINE
    metadata: Dict[str, Any] = field(default_factory=dict)
    last_heartbeat: float = 0.0
    created_at: float = field(default_factory=time.time)


class OFAAgent:
    """OFA Python Agent"""

    def __init__(self, config: Optional[AgentConfig] = None):
        self.config = config or AgentConfig()
        self.info = AgentInfo(
            id=self.config.agent_id,
            name=self.config.name,
            skills=self.config.skills,
            metadata=self.config.metadata,
        )
        self.state = AgentState.INITIALIZING
        self._connection: Optional[Any] = None
        self._skill_registry: Dict[str, Callable] = {}
        self._task_handlers: Dict[str, Callable] = {}
        self._running = False
        self._heartbeat_task: Optional[asyncio.Task] = None
        self._message_handlers: Dict[str, List[Callable]] = {}

        logger.info(f"OFA Agent initialized: {self.info.id}")

    async def connect(self) -> bool:
        """连接到Center"""
        self.state = AgentState.CONNECTING
        logger.info(f"Connecting to {self.config.center_url}")

        try:
            # 根据连接类型创建连接
            if self.config.connection_type == "grpc":
                from .connection import GRPCConnection
                self._connection = GRPCConnection(self.config)
            elif self.config.connection_type == "http":
                from .connection import HTTPConnection
                self._connection = HTTPConnection(self.config)
            elif self.config.connection_type == "websocket":
                from .connection import WebSocketConnection
                self._connection = WebSocketConnection(self.config)

            await self._connection.connect()
            await self._register()
            self.state = AgentState.ONLINE
            self._running = True

            # 启动心跳
            self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())

            logger.info(f"Agent connected: {self.info.id}")
            return True

        except Exception as e:
            self.state = AgentState.ERROR
            logger.error(f"Connection failed: {e}")
            return False

    async def disconnect(self) -> None:
        """断开连接"""
        self._running = False

        if self._heartbeat_task:
            self._heartbeat_task.cancel()
            try:
                await self._heartbeat_task
            except asyncio.CancelledError:
                pass

        if self._connection:
            await self._connection.disconnect()

        self.state = AgentState.OFFLINE
        logger.info(f"Agent disconnected: {self.info.id}")

    async def _register(self) -> None:
        """注册到Center"""
        register_msg = {
            "type": "register",
            "agent_id": self.info.id,
            "name": self.info.name,
            "type": self.info.type,
            "version": self.info.version,
            "platform": self.info.platform,
            "skills": self.info.skills,
            "metadata": self.info.metadata,
        }
        await self._connection.send(register_msg)

    async def _heartbeat_loop(self) -> None:
        """心跳循环"""
        while self._running:
            try:
                heartbeat_msg = {
                    "type": "heartbeat",
                    "agent_id": self.info.id,
                    "state": self.state.value,
                    "timestamp": time.time(),
                }
                await self._connection.send(heartbeat_msg)
                self.info.last_heartbeat = time.time()
                await asyncio.sleep(self.config.heartbeat_interval)
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Heartbeat failed: {e}")
                await self._reconnect()

    async def _reconnect(self) -> None:
        """重连"""
        for attempt in range(self.config.max_reconnect_attempts):
            self.state = AgentState.CONNECTING
            logger.info(f"Reconnecting (attempt {attempt + 1})")

            try:
                await self._connection.connect()
                await self._register()
                self.state = AgentState.ONLINE
                self._running = True
                logger.info("Reconnected successfully")
                return
            except Exception as e:
                logger.error(f"Reconnect failed: {e}")
                await asyncio.sleep(self.config.reconnect_interval)

        self.state = AgentState.ERROR
        logger.error("Max reconnect attempts reached")

    def register_skill(self, skill_id: str, handler: Callable) -> None:
        """注册技能"""
        self._skill_registry[skill_id] = handler
        self.info.skills.append(skill_id)
        logger.info(f"Skill registered: {skill_id}")

    def unregister_skill(self, skill_id: str) -> None:
        """注销技能"""
        self._skill_registry.pop(skill_id, None)
        self.info.skills = [s for s in self.info.skills if s != skill_id]

    def on_message(self, msg_type: str, handler: Callable) -> None:
        """注册消息处理器"""
        if msg_type not in self._message_handlers:
            self._message_handlers[msg_type] = []
        self._message_handlers[msg_type].append(handler)

    async def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """执行任务"""
        skill_id = task.get("skill_id")
        operation = task.get("operation")
        input_data = task.get("input", {})

        self.state = AgentState.BUSY

        try:
            handler = self._skill_registry.get(skill_id)
            if not handler:
                raise ValueError(f"Skill not found: {skill_id}")

            # 支持同步和异步处理器
            if asyncio.iscoroutinefunction(handler):
                result = await handler(operation, input_data)
            else:
                result = handler(operation, input_data)

            self.state = AgentState.ONLINE
            return {
                "success": True,
                "result": result,
                "agent_id": self.info.id,
            }

        except Exception as e:
            self.state = AgentState.ERROR
            logger.error(f"Task execution failed: {e}")
            return {
                "success": False,
                "error": str(e),
                "agent_id": self.info.id,
            }

    async def send_message(self, target: str, msg_type: str, data: Any) -> None:
        """发送消息"""
        msg = {
            "type": "message",
            "from": self.info.id,
            "to": target,
            "msg_type": msg_type,
            "data": data,
            "timestamp": time.time(),
        }
        await self._connection.send(msg)

    async def broadcast(self, msg_type: str, data: Any) -> None:
        """广播消息"""
        msg = {
            "type": "broadcast",
            "from": self.info.id,
            "msg_type": msg_type,
            "data": data,
            "timestamp": time.time(),
        }
        await self._connection.send(msg)

    async def run(self) -> None:
        """运行Agent"""
        await self.connect()

        try:
            while self._running:
                msg = await self._connection.receive()
                if msg:
                    await self._handle_message(msg)
        except asyncio.CancelledError:
            pass
        finally:
            await self.disconnect()

    async def _handle_message(self, msg: Dict[str, Any]) -> None:
        """处理消息"""
        msg_type = msg.get("type")

        # 任务消息
        if msg_type == "task":
            result = await self.execute_task(msg)
            await self._connection.send({
                "type": "task_result",
                "task_id": msg.get("task_id"),
                **result,
            })

        # 其他消息
        handlers = self._message_handlers.get(msg_type, [])
        for handler in handlers:
            try:
                if asyncio.iscoroutinefunction(handler):
                    await handler(msg)
                else:
                    handler(msg)
            except Exception as e:
                logger.error(f"Message handler failed: {e}")

    @property
    def is_online(self) -> bool:
        """是否在线"""
        return self.state == AgentState.ONLINE

    def get_info(self) -> Dict[str, Any]:
        """获取Agent信息"""
        return {
            "id": self.info.id,
            "name": self.info.name,
            "type": self.info.type,
            "version": self.info.version,
            "platform": self.info.platform,
            "skills": self.info.skills,
            "state": self.state.value,
            "metadata": self.info.metadata,
            "last_heartbeat": self.info.last_heartbeat,
        }


async def create_agent(config: Optional[AgentConfig] = None) -> OFAAgent:
    """创建Agent实例"""
    agent = OFAAgent(config)
    return agent


def run_agent(config: Optional[AgentConfig] = None) -> None:
    """运行Agent（同步入口）"""
    agent = OFAAgent(config)
    asyncio.run(agent.run())

File: src/sdk/nodejs/ofa-agent/src/agent.ts
/**
 * Agent Core Module
 * Sprint 29: Node.js Agent SDK
 */

import { v4 as uuidv4 } from 'uuid';
import { AgentConfig, DefaultAgentConfig } from './config';
import { Connection, createConnection } from './connection';
import { SkillExecutor, SkillRegistry } from './skills';
import { Message, MessageType, Task, TaskResult } from './message';

/**
 * Agent状态
 */
export enum AgentState {
  INITIALIZING = 'initializing',
  CONNECTING = 'connecting',
  ONLINE = 'online',
  BUSY = 'busy',
  OFFLINE = 'offline',
  ERROR = 'error',
}

/**
 * Agent信息
 */
export interface AgentInfo {
  id: string;
  name: string;
  type: string;
  version: string;
  platform: string;
  skills: string[];
  state: AgentState;
  metadata: Record<string, unknown>;
  lastHeartbeat: number;
}

/**
 * Agent统计
 */
export interface AgentStats {
  tasksExecuted: number;
  tasksSuccess: number;
  tasksFailed: number;
  messagesSent: number;
  messagesReceived: number;
}

/**
 * OFA Agent
 */
export class OFAAgent {
  private config: AgentConfig;
  private _state: AgentState = AgentState.INITIALIZING;
  private _connection: Connection | null = null;
  private _skillExecutor: SkillExecutor;
  private _skillRegistry: SkillRegistry;
  private _stats: AgentStats;
  private _running: boolean = false;
  private _heartbeatTimer?: NodeJS.Timeout;
  private _messageHandlers: Map<string, Set<(msg: Message) => void>> = new Map();

  /**
   * Agent ID
   */
  public readonly id: string;

  /**
   * Agent信息
   */
  public readonly info: AgentInfo;

  constructor(config?: Partial<AgentConfig>) {
    this.config = { ...DefaultAgentConfig(), ...config };
    this.id = this.config.agentId || `nodejs-agent-${uuidv4().split('-')[0]}`;

    this.info = {
      id: this.id,
      name: this.config.name,
      type: 'nodejs',
      version: '8.1.0',
      platform: 'nodejs',
      skills: [],
      state: AgentState.INITIALIZING,
      metadata: {},
      lastHeartbeat: 0,
    };

    this._skillExecutor = new SkillExecutor();
    this._skillRegistry = new SkillRegistry();
    this._stats = {
      tasksExecuted: 0,
      tasksSuccess: 0,
      tasksFailed: 0,
      messagesSent: 0,
      messagesReceived: 0,
    };
  }

  /**
   * 获取当前状态
   */
  get state(): AgentState {
    return this._state;
  }

  /**
   * 是否在线
   */
  get isOnline(): boolean {
    return this._state === AgentState.ONLINE;
  }

  /**
   * 连接到Center
   */
  async connect(): Promise<boolean> {
    this._state = AgentState.CONNECTING;
    console.log(`Connecting to ${this.config.centerUrl}`);

    try {
      this._connection = createConnection(this.config);
      await this._connection.connect();
      await this._register();
      this._state = AgentState.ONLINE;
      this._running = true;
      this._startHeartbeat();

      console.log(`Agent connected: ${this.id}`);
      return true;
    } catch (error) {
      this._state = AgentState.ERROR;
      console.error('Connection failed:', error);
      return false;
    }
  }

  /**
   * 断开连接
   */
  async disconnect(): Promise<void> {
    this._running = false;

    if (this._heartbeatTimer) {
      clearInterval(this._heartbeatTimer);
    }

    if (this._connection) {
      await this._connection.disconnect();
    }

    this._state = AgentState.OFFLINE;
    console.log(`Agent disconnected: ${this.id}`);
  }

  /**
   * 注册到Center
   */
  private async _register(): Promise<void> {
    const msg: Message = {
      id: uuidv4(),
      type: MessageType.REGISTER,
      from: this.id,
      to: '',
      subject: '',
      data: { ...this.info },
      timestamp: Date.now(),
    };
    await this._connection!.send(msg);
  }

  /**
   * 启动心跳
   */
  private _startHeartbeat(): void {
    this._heartbeatTimer = setInterval(async () => {
      if (!this._running || !this._connection) return;

      const msg: Message = {
        id: uuidv4(),
        type: MessageType.HEARTBEAT,
        from: this.id,
        to: '',
        subject: '',
        data: { state: this._state },
        timestamp: Date.now(),
      };

      try {
        await this._connection.send(msg);
        this.info.lastHeartbeat = Date.now();
      } catch (error) {
        console.error('Heartbeat failed:', error);
        await this._reconnect();
      }
    }, this.config.heartbeatInterval * 1000);
  }

  /**
   * 重连
   */
  private async _reconnect(): Promise<void> {
    for (let i = 0; i < this.config.maxReconnectAttempts; i++) {
      this._state = AgentState.CONNECTING;
      console.log(`Reconnecting (attempt ${i + 1})`);

      try {
        await this._connection?.connect();
        await this._register();
        this._state = AgentState.ONLINE;
        this._running = true;
        console.log('Reconnected successfully');
        return;
      } catch (error) {
        console.error('Reconnect failed:', error);
        await new Promise(r => setTimeout(r, this.config.reconnectInterval * 1000));
      }
    }

    this._state = AgentState.ERROR;
    console.error('Max reconnect attempts reached');
  }

  /**
   * 注册技能
   */
  registerSkill(skillId: string, handler: (operation: string, input: unknown) => Promise<unknown>): void {
    this._skillExecutor.register(skillId, handler);
    this.info.skills.push(skillId);
    console.log(`Skill registered: ${skillId}`);
  }

  /**
   * 注销技能
   */
  unregisterSkill(skillId: string): void {
    this._skillExecutor.unregister(skillId);
    this.info.skills = this.info.skills.filter(s => s !== skillId);
  }

  /**
   * 执行任务
   */
  async executeTask(task: Task): Promise<TaskResult> {
    this._state = AgentState.BUSY;

    const start = Date.now();
    let result: TaskResult;

    try {
      const output = await this._skillExecutor.execute(task.skillId, task.operation, task.input);
      result = {
        taskId: task.taskId,
        success: true,
        result: output,
        agentId: this.id,
        durationMs: Date.now() - start,
      };
      this._stats.tasksSuccess++;
    } catch (error) {
      result = {
        taskId: task.taskId,
        success: false,
        error: String(error),
        agentId: this.id,
        durationMs: Date.now() - start,
      };
      this._stats.tasksFailed++;
    }

    this._stats.tasksExecuted++;
    this._state = AgentState.ONLINE;
    return result;
  }

  /**
   * 发送消息
   */
  async sendMessage(target: string, type: MessageType, data: unknown): Promise<void> {
    if (!this._connection) throw new Error('Not connected');

    const msg: Message = {
      id: uuidv4(),
      type,
      from: this.id,
      to: target,
      subject: '',
      data,
      timestamp: Date.now(),
    };

    await this._connection.send(msg);
    this._stats.messagesSent++;
  }

  /**
   * 广播消息
   */
  async broadcast(type: MessageType, data: unknown): Promise<void> {
    await this.sendMessage('', type, data);
  }

  /**
   * 注册消息处理器
   */
  onMessage(type: string, handler: (msg: Message) => void): void {
    if (!this._messageHandlers.has(type)) {
      this._messageHandlers.set(type, new Set());
    }
    this._messageHandlers.get(type)!.add(handler);
  }

  /**
   * 运行Agent
   */
  async run(): Promise<void> {
    await this.connect();

    while (this._running && this._connection) {
      try {
        const msg = await this._connection.receive();
        if (msg) {
          await this._handleMessage(msg);
          this._stats.messagesReceived++;
        }
      } catch (error) {
        console.error('Message handling error:', error);
      }

      await new Promise(r => setTimeout(r, 100));
    }
  }

  /**
   * 处理消息
   */
  private async _handleMessage(msg: Message): Promise<void> {
    if (msg.type === MessageType.TASK) {
      const task = msg.data as Task;
      const result = await this.executeTask(task);

      await this._connection?.send({
        id: uuidv4(),
        type: MessageType.TASK_RESULT,
        from: this.id,
        to: '',
        subject: '',
        data: result,
        timestamp: Date.now(),
      });
    }

    // 调用自定义处理器
    const handlers = this._messageHandlers.get(msg.type);
    if (handlers) {
      for (const handler of handlers) {
        try {
          handler(msg);
        } catch (error) {
          console.error('Handler error:', error);
        }
      }
    }
  }

  /**
   * 获取统计
   */
  getStats(): AgentStats {
    return { ...this._stats };
  }
}

// 导出配置
export { AgentConfig, DefaultAgentConfig } from './config';